From 5fddc7a2763d65d658baf89fd6ddabba5d4e91f4 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Tue, 19 Aug 2014 19:30:52 -0400
Subject: [scsi] qla4xxx: Fix memory leak for ha->saved_acb

Message-id: <1408476658-5056-15-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88353
O-Subject: [PATCH RHEL 7.1 e-stor 14/20] qla4xxx: Fix memory leak for ha->saved_acb
Bugzilla: 1089349
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Nilesh Javali <nilesh.javali@qlogic.com>

scsi commit id f65241e7294e344f4641131000ea88e3d70fa550

Signed-off-by: Vikas Chaudhary <vikas.chaudhary@qlogic.com>
Reviewed-by: Mike Christie <michaelc@cs.wisc.edu>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla4xxx/ql4_mbx.c b/drivers/scsi/qla4xxx/ql4_mbx.c
index 1345c0e..0a3312c 100644
--- a/drivers/scsi/qla4xxx/ql4_mbx.c
+++ b/drivers/scsi/qla4xxx/ql4_mbx.c
@@ -2395,8 +2395,6 @@ int qla4_84xx_config_acb(struct scsi_qla_host *ha, int acb_config)
   }
 
   memcpy(acb, ha->saved_acb, acb_len);
-  kfree(ha->saved_acb);
-  ha->saved_acb = NULL;
 
   rval = qla4xxx_set_acb(ha, &mbox_cmd[0], &mbox_sts[0], acb_dma);
   if (rval != QLA_SUCCESS)
@@ -2412,6 +2410,10 @@ exit_free_acb:
  dma_free_coherent(&ha->pdev->dev, sizeof(struct addr_ctrl_blk), acb,
      acb_dma);
 exit_config_acb:
+ if ((acb_config == ACB_CONFIG_SET) && ha->saved_acb) {
+  kfree(ha->saved_acb);
+  ha->saved_acb = NULL;
+ }
  DEBUG2(ql4_printk(KERN_INFO, ha,
      "%s %s\n", __func__,
      rval == QLA_SUCCESS ? "SUCCEEDED" : "FAILED"));
-- 
1.7.1